### DISCUSSION TITLE
How to run bot and flask

### DISCUSSION THREAD

**Discussion by JhonathaNunes** [python-telegram-bot/python-telegram-bot#2663] [Q&A] [answered]
Hello there. I am not used to python and Flask, but I have to use it for a project. I have to make  flask app and integrate it with telegram somehow. My flask app is ok and I made a example bot and if I run it alone it works fine, but I wanted to know how can I run it alongside with my flask app. Here is what I tried, but there is a conflict exception:

```
from config import Config
from telegram import Update
from telegram.ext import (
    Updater,
    CommandHandler,
    MessageHandler,
    Filters,
    CallbackContext
)
from app import init_app

app = init_app()


def start(update: Update, context: CallbackContext):
    update.message.reply_text("Oooopa")


def echo_reverse(update: Update, context: CallbackContext) -> None:
    update.message.reply_text(update.message.text[::1])


def main() -> None:
    updater = Updater(Config.TELEGRAM_TOKEN)

    dispatcher = updater.dispatcher

    dispatcher.add_handler(CommandHandler("start", start))

    dispatcher.add_handler(MessageHandler(
        Filters.text & ~Filters.command,
        echo_reverse
    ))

    updater.start_polling()
    app.run(host="0.0.0.0", debug=True)
    updater.idle()
```
I also tried to run the flask app first, but then only the flask app starts, the bot just works after I ctrl + c and if I run the app after updater.idle() I have to stop the bot so my flask app runs.

Sorry if I am being stupid or something, but I wanted to understand and know how to solve my problem

**Comment by Bibo-Joshi** [comment, marked as answer]
I guess you'd have to run either `updater.start_polling` or `app.run` in a different thread. Alternatively, you could switch to getting updates via webhooks and use flask to fetch them instead of `updater`. see the wiki page on [webhooks](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Webhooks) for more details on that.

**Reply by JhonathaNunes** [reply]
Hi there, just to update. I made a different thread to run flask and it works just fine. I just had to remove the debug flag since it just listens to updates at the main thread.
```
from app import init_app
from app.telegram_bot.telegram_bot import main
import threading
app = init_app()


class FlaskThread(threading.Thread):
    def run(self) -> None:
        app.run(host="0.0.0.0")


class TelegramThread(threading.Thread):
    def run(self) -> None:
        main()


if __name__ == "__main__":
    flask_thread = FlaskThread()
    flask_thread.start()

    main()
```

**Reply by iSarabjitDhiman** [reply]
Quick fix: Make sure to set the background running thread as a Daemon thread, so that it gets terminated as soon as the main thread stops.
With the current code, if u press Ctrl + C, the main thread i.e. main() gets terminated but flask_thread stays alive in the background. Just set the thread.daemon = True and it fixes the problem.

Now the new code will look similar to this one:

```python

from app import init_app
from app.telegram_bot.telegram_bot import main
import threading
app = init_app()


class FlaskThread(threading.Thread):
    def run(self) -> None:
        app.run(host="0.0.0.0")


class TelegramThread(threading.Thread):
    def run(self) -> None:
        main()


if __name__ == "__main__":
    flask_thread = FlaskThread()
    flask_thread.daemon = True
    flask_thread.start()

    main()
```
